### ISSUE TITLE
UDP amplification?

### ISSUE THREAD

**Issue by riastradh** [heimdal/heimdal#1216]
When I run kinit with an existing principal but the wrong password in a realm I use, it sends a 175-byte UDP packet to the KDC, and the KDC replies with a 313- or 367-byte packet.  (kinit and KDC are both Heimdal.)

I haven't looked at the details of the messages here but evidently authentication is not required to provoke this ~2x-sized reply from the KDC, so it appears to serve as a DoS amplification vector.  Perhaps not as big as a DNS ANY or DNSSEC query reply, but still ~2x.

Does Heimdal do anything to avoid or mitigate this (preferably out of the box), other than disabling UDP in the KDC?

**Comment by nicowilliams** [comment]
Hmm, w/o FAST armoring I'm observing a 160 byte response to a 289 byte request.  However, PKINIT can probably be made to produce some amplification.

I believe Heimdal does nothing about this.  In the open Internet it'd be best to disable UDP or filter it out.  If this is a problem we could implement a throttle that sends back reduced `KRB-ERROR` responses, or no responses when requests fail.  Because the Heimdal KDC is multi-processed we'd have to keep a per-process throttle, an exponentially decaying running average of failing request rate, and trigger the mitigation on some level.

**Comment by nicowilliams** [comment]
> Can UDP access be limited to using a TGT to get a service ticket? And if that path fails because the TGT is bad, does it amplify?

In principle the AS and TGS can be separate services.  In practice the discovery mechanisms don't support that... but the AS could respond with a bare-bones (all `OPTIONAL` fields left out) `KRB-ERROR` w/ `KRB_ERR_FIELD_TOOLONG`.

**Comment by elric1** [comment]
You don't have to entirely disable UDP, necessarily, you could always send `KRB5KRB_ERR_RESPONSE_TOO_BIG` if certain amplification criteria are met, that is: if the response packet is more than a certain percentage larger than the input packet.
I don't think that it makes sense to directly distinguish between AS and TGS, because that's not the determinant of whether it's an amplification: the size of the response packet is. IIRC, there is already logic in the code to deal with responses that are too large, so it shouldn't be terribly hard to implement.

**Comment by nicowilliams** [comment]
Oops, yes, I meant that error, not the other.

**Comment by nicowilliams** [comment]
I'd be inclined to do this for all AS requests, really, because the first one always fails as its purpose is to discover the pre-authentication options available.  As long as it interops, I'd be fine with that.

**Comment by elric1** [comment]
Might be an idea to have the client prefer TCP for the AS in all cases if we do this.
It might be possible to get larger amplifications using the contents of the PAC. Certainly with a Windows KDC, an AS_REQ can potentially return a very large TGT for some users. Granted, IIRC, Windows only uses TCP, we may support large PACs either now or in a potential future release.
This can be a potential problem for the TGS as well as, I think, that ticket renewal, which uses the TGS, may want to re-evaluate group membership to construct a new PAC, or, perhaps, x-realm TGS TGT requests might also want to reconstruct PACs. I haven't looked into this, but it might be worth considering.
Do we do anything with PACs at the moment?

**Comment by nicowilliams** [comment]
Oh, and remember that the KDC protocol is stateless, so replays within the skew window are a way to generate amplification in both the AS _and_ TGS cases.  So I think maybe we do need an adaptive mechanism, and it shouldn't only be driven by failures because the fact that the request is authenticated doesn't prove it's not a replay.

**Comment by nicowilliams** [comment]
> Do we do anything with PACs at the moment?

Yes, but very limited (no group membership list, so our PACs are always small).

**Comment by nicowilliams** [comment]
Maybe we want to return `KRB5KRB_ERR_RESPONSE_TOO_BIG` whenever the reply is larger than the request, or even always, and maybe we want the client to try TCP first.  I'd be happy to go this route.

**Comment by nicowilliams** [comment]
There is the `[kdc]` `max-kdc-datagram-reply-length` parameter, which defaults to 1400 (reply PDU size in bytes), which causes Heimdal to send `KRB5_ERR_RESPONSE_TOO_BIG` when reply PDU sizes exceed that.  So there is a configuration parameter that is available now that can be tuned way down to force the use of TCP.

**Comment by nicowilliams** [comment]
> It looks like there is logic to apply a statically configured limit to AS-REP packets over UDP, but I can't find it for TGS-REP packets:

```
: ; git grep max_datagram
ChangeLog.2006: * kdc/kdc.h (krb5_kdc_config): Add max_datagram_reply_length.
kdc/default_config.c:    c->max_datagram_reply_length =
kdc/kdc_locl.h:    size_t max_datagram_reply_length;
kdc/kerberos5.c:    if (r->datagram_reply && r->reply->length > config->max_datagram_reply_length) {
kdc/krb5tgs.c:    if (datagram_reply && data->length > config->max_datagram_reply_length) {
```

**Comment by nicowilliams** [comment]
> What's the smallest AS-REQ or TGS-REQ packet that Heimdal will reply to?

It depends on the principal and realm names, but ~200 bytes.

If one uses PKINIT it's larger then the reply can easily be larger than the 1400 byte default UDP PDU limit.  With ECC though it's possible that PKINIT messages will fit in 1400 bytes.  I'm seeing almost 3x reply size with PKINIT.

> How large is a KRB5_ERR_RESPONSE_TOO_BIG KRB-ERROR packet?

Less than 200 bytes w/o FAST, around 600 bytes w/ FAST.  When using FAST error responses are generally, if not always smaller than the request.

**Comment by nicowilliams** [comment]
Note that the `[kdc]` `max-kdc-datagram-reply-length` parameter does not apply to the `KRB5_ERR_RESPONSE_TOO_BIG` error PDU itself!  Therefore we can easily and safely (as far as interop goes) tune it down to zero.

**Comment by nicowilliams** [comment]
It isn't currently possible to disable listening on UDP, but that's easily handled with IP filtering anyways.

**Comment by nicowilliams** [comment]
> My experiment with riastradh@NETBSD.ORG was only 175 bytes. Can we write down an exact lower bound to make a proof of non-amplification easier?

One would have to either write [a small amount of] code or set up a 1 character realm with a 1 character principal, or calculate it by hand from the ASN.1 types and knowledge of DER.  If I'd written a JSON->DER converter to go with the DER->JSON converter that `lib/asn1` currently has then it'd be easier :/ (this is something I eventually want to do, but there's a huge pile of other work to do on lots of things).

**Comment by nicowilliams** [comment]
> If smallest request that can provoke a reply is n bytes, and if a KRB5_ERR_RESPONSE_TOO_BIG reply is at most m bytes, then all we have to prove is m <= n to be done with it once Heimdal replaces any longer-than-request replies by KRB5_ERR_RESPONSE_TOO_BIG.

My take is that we can solve this with a) advice to sites w/ KDCs in the open to not advertise DNS `SRV` RRs for UDP, c, advice to sites to use IP filtering blackhole (or reject) UDP requests to port 88 on their KDCs in the opem, c) us setting the `max-kdc-datagram-reply-length` default to zero.  Sure, (c) doesn't prevent the attack if the smallest `AS-REQ` could be much smaller than the smallest `KRB-ERROR`, though I don't think that's likely.

Also, we should make sure that Heimdal's `KRB5_ERR_RESPONSE_TOO_BIG` replies are the smallest possible by dropping all non-essential `OPTIONAL` fields from the reply.  This won't help that much in the FAST case, but in the FAST case the error is practically guaranteed to be smaller than the request anyways.

**Comment by nicowilliams** [comment]
Although RFC 6113 doesn't say, `KRB5_ERR_RESPONSE_TOO_BIG` effectively has to always be sent as a non-FAST `KRB-ERROR`, even when the client used FAST.

**Comment by nicowilliams** [comment]
> Should this have a CVE assigned and mitigation instructions for Kerberos administrators? (Am I the first person to have thought of this? Surely not?)

IMO it's of low urgency because there are much better amplification attacks out there.  But I could be convinced otherwise.  A bigger problem is that this wouldn't be specific to Heimdal -- this is a problem with an Internet protocol, so the next step would be to seek to publish an update to RFC 4120, or a BCP (Best Current Practice) perhaps, but who will do it?  We'd also need to reach out to @greghudson (MIT) and @SteveSyfuhs (MSFT) to get their take on this.

**Comment by SteveSyfuhs** [comment]
We (Windows) considered this kind of attack a low severity issue. Amplification attacks in corporate networks just aren't a thing, and on top of that the amount of requests required to saturate a standard Gbit LAN is pretty high. Network ops folks would catch that relatively quickly. It's just not the same environment as the open internet.
________________________________

**Comment by nicowilliams** [comment]
Perhaps the best fix is to a) not advertise UDP in DNS (nor local `krb5.conf`s) and b) not serve Kerberos on UDP.  I.e., this is just a docs fix.

**Comment by nicowilliams** [comment]
> > Perhaps the best fix is to a) not advertise UDP in DNS (nor local `krb5.conf`s)
> 
> Well, what clients try first is just a matter of usability and performance, and, if TCP has been disabled, compatibility. Changing normal client behaviour won't affect the real issue, which is how the KDC responds when provoked.

Not really.  Clients generally try UDP first when it's advertised, thus not advertising it is a way to get clients not to try UDP.  Naturally that's not enough IF the KDC nonetheless would respond to requests sent via UDP, thus the `and` that followed this.  To disable UDP on the KDC but continue advertising it would only serve to cause annoying timeouts for clients.

> > and b) not serve Kerberos on UDP. I.e., this is just a docs fix.
> 
> Sure, that's reasonable. I think it would be best if the KDC had some simple reliable default-enabled mechanism to avoid returning UDP replies that are longer than the requests that provoked them. But at least clear suggestions about how to prevent UDP replies would be good (which was the line of thought that prompted me to file #1223).

I've written such a patch, but the best fix is no code, because adding code risks adding bugs and adds to the future maintenance burden.  That said, it will be a lot easier to get this "problem" (if it really is one) to go away if we patch it than if we tell operators how to configure it away.  It's just not clear to me that this really is a serious problem.

**Comment by nicowilliams** [comment]
@SteveSyfuhs does the Windows client accept non-FAST `KRB-ERROR` w/ `error-code` set to `KRB_ERR_RESPONSE_TOO_BIG` when the request sent had been a FAST request?

And does the AD KDC respond with non-FAST `KRB-ERROR` w/ `error-code` set to `KRB_ERR_RESPONSE_TOO_BIG` when the response would be too big, or does it send the `KRB_ERR_RESPONSE_TOO_BIG` in an inner `KRB-ERROR`?

**Comment by SteveSyfuhs** [comment]
I think we just set the standard error with TOO-BIG. I don't think we can guarantee a FAST error will fit in a UDP message.
________________________________

**Comment by nicowilliams** [comment]
> I think we just set the standard error with TOO-BIG. I don't think we can guarantee a FAST error will fit in a UDP message.

Thanks.  I assume that means in the outer `KRB-ERROR` (and no inner one).  RFC 6113 kinda says to do just that too, and MIT does that.  So I'll make sure Heimdal does too.  MIT and Heimdal only correctly process the `TOO-BIG` error when it's on the outer `KRB-ERROR`, FYI.

**Comment by riastradh** [comment]
In answer to my earlier question, I think this is the smallest possible KDC-REQ (AS-REQ or TGS-REQ) packet, unless there is some variation allowed in ASN.1 encoding which Heimdal's encoder doesn't implement:

```
00000000  6a 29 30 27 a1 03 02 01  05 a2 03 02 01 0a a4 1b  |j)0'............|
00000010  30 19 a0 07 03 05 00 00  00 00 00 a2 02 1b 00 a7  |0...............|
00000020  06 02 04 12 34 56 78 a8  02 30 00                 |....4Vx..0.|
```

This 51-byte packet provokes a 112-byte response, for just over 2x amplification.  That's not an upper or lower bound on the amplification factor, of course---perhaps a _slightly_ larger KDC-REQ packet can provoke a _much_ larger KDC-REP.

For kpasswdd(8), `00 06 08 ff 00 06` gives me a 6-to-126-byte amplification factor, or 21x.  But because of https://github.com/heimdal/heimdal/issues/1307, it can't simply be switched to TCP-only, unlike kdc(8)/AS/TGS.

---

Minimal KDC-REQ packet produced by the following program:

```
#include <err.h>
#include <errno.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <krb5/krb5.h>
#include <krb5/krb5_asn1.h>

int
main(void)
{
	krb5_data req;
	AS_REQ as;
	size_t len = 0;
	int error;

	memset(&as, 0, sizeof(as));

	as.pvno = 5;
	as.msg_type = krb_as_req;
	as.req_body.realm = "";
	as.req_body.nonce = 0x12345678;

	ASN1_MALLOC_ENCODE(AS_REQ, req.data, req.length, &as, &len, error);
	if (error)
		errx(EXIT_FAILURE, "ASN1_MALLOC_ENCODE: %d", error);
	if (fwrite(req.data, req.length, 1, stdout) != 1)
		err(EXIT_FAILURE, "fwrite");
	if (fflush(stdout) == EOF)
		err(EXIT_FAILURE, "fflush");
	return ferror(stdout);
}
```

**Comment by nicowilliams** [comment]
@riastradh 

> perhaps a slightly larger KDC-REQ packet can provoke a much larger KDC-REP.

That's always possible if you have KDC plugins to add AD-style PACs to issued tickets, and all you have to do is replay `TGS-REQ`s that have a valid `PA-TGS`, and you have 5 minutes to do it in.  You can get huge amplification rates this way, but of course you want not to exceed the KDC's UDP size limit.  This affects every existing TGS.

Ultimately the answer here is to just tune down the max UDP response size so the KDC responds with `TOO-BIG`.

Regarding `kpasswdd`, RFC 3244 does define a TCP protocol on port 464, and we can teach `lib/krb5/changepw.c`/`lib/krb5/krbhst.c`/`kpasswd/kpasswdd.c` to use the TCP protocol preferentially and fall back on UDP when `ECONNREFUSED`.

**Comment by nicowilliams** [comment]
#1218 -- @riastradh or @elric1, can you review?
